README.md
<h1 align="center">Welcome to Gotify Indicator 👋</h1>

![Licencia MIT](https://img.shields.io/badge/Licencia-MIT-green)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/b3e704c3f150404582cd23b9fcb4be32)](https://www.codacy.com/manual/atareao/Gotify-Indicator?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=atareao/gotify-indicator&amp;utm_campaign=Badge_Grade)
[![CodeFactor](https://www.codefactor.io/repository/github/atareao/gotify-indicator/badge/master)](https://www.codefactor.io/repository/github/atareao/gotify-indicator/overview/master)

[![Twitter: atareao](https://img.shields.io/twitter/follow/atareao.svg?style=social)](https://twitter.com/atareao)

Gotify Indicator is an application to recieve and send messages from your own Gotify Server.

[![Gotify-Indicator](./data/icons/gotify-indicator.svg)](https://www.atareao.es/aplicacion/gotify-indicator/)

## 🏠 [Homepage](https://www.atareao.es/applicacion/gotify-indicator)

## Prerequisites

Before you begin, ensure you have met the following requirements:

* If you install it from PPA don't worry about, becouse all the requirements are included in the package
* If you clone the repository, you need, at least, these dependecies,

```
python3
python3-pydub
gir1.2-gtk-3.0
gir1.2-glib-2.0
gir1.2-gdkpixbuf-2.0
gir1.2-appindicator3-0.1
gir1.2-notify-0.7
```

## Installing Gotify-Indicator

To install **Gotify-Indicator**, follow these steps:

* In a terminal (`Ctrl+Alt+T`), run these commands

```
sudo add-apt-repository ppa:atareao/atareao
sudo apt update
sudo apt install gotify-indicator
```

## Using Gotify Indicator

To use **Gotify Indicator**, open it, and configure it. First, create an application in the Gotify web. Write down, application name and application token.

![Application Configuration](./screenshots/aplications.png)

Create a client in the Gotify web. Write down, client token

![Client Configuration](./screenshots/clients.png)

In the configuration of Gotify Indicator set the token for the application and the client

![Gotify-Indicator Configuration](./screenshots/configuracion.png)

## Contributing to Gotify Indicator

To contribute to **Gotify Indicator**, follow these steps:

1. Fork this repository.
2. Create a branch: `git checkout -b <branch_name>`.
3. Make your changes and commit them: `git commit -m '<commit_message>'`
4. Push to the original branch: `git push origin <project_name>/<location>`
5. Create the pull request.

Alternatively see the GitHub documentation on [creating a pull request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request).

## 👤 Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://github.com/albanobattistella"><img src="https://avatars.githubusercontent.com/u/34811668" width="100px;" alt=""/><br /><sub><b>albanobattistella</b></sub></a><br /><a href="https://github.com/atareao/gotify-indicator/commits?author=albanobattistella" title="Code">💻</a></td>
    <td align="center"><a href="https://www.atareao.es"><img src="https://avatars3.githubusercontent.com/u/298055?v=4" width="100px;" alt=""/><br /><sub><b>Lorenzo Carbonell</b></sub></a><br /><a href="https://github.com/atareao/fondos-productivos/commits?author=atareao" title="Code">💻</a></td>
    <td align="center"><a href="https://github.com/newhinton"><img src="https://avatars.githubusercontent.com/u/25279821?" width="100px;" alt=""/><br /><sub><b>newhinton</b></sub></a><br /><a href="https://github.com/atareao/gotify-indicator/commits?author=newhinton" title="Code">💻</a></td>
  </tr>
</table>

<!-- markdownlint-enable -->
<!-- prettier-ignore-end -->
<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!

## Contact

If you want to contact me you can reach me at [atareao.es](https://www.atareao.es).

## License

This project uses the following license: [MIT License](https://choosealicense.com/licenses/mit/).


debian/changelog
gotify-indicator (0.2.0-0extras20.04.0) focal; urgency=medium

  * Add imagecaching
  * Add notification sound

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 17 Apr 2021 20:53:23 +0200

gotify-indicator (0.1.1-0extras18.04.3) bionic; urgency=medium

  * Check if there is configuration
  * Add spanish translation

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Tue, 04 Feb 2020 21:05:56 +0100

gotify-indicator (0.1.0-0extras18.04.03) bionic; urgency=medium

  * First version

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Thu, 30 Jan 2020 19:49:17 +0100


src/cache.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# This file is part of gotify-indicator
#
# Copyright (c) 2021 Felix Nüsse
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
import os

from configurator import Configuration
from config import CACHE_DIR
import requests
import threading
import shutil


class Cache(threading.Thread):

    def __init__(self, https_protocol, base_url, client_token):
        threading.Thread.__init__(self)
        protocol = 'https' if https_protocol else 'http'
        self.url = '{}://{}'.format(protocol, base_url)
        self.client_token = client_token

    def startCaching(self):
        headers = {'X-Gotify-Key': self.client_token}
        url = '{}/{}'.format(self.url, 'application')
        ans = requests.get(url, headers=headers)
        for app in ans.json():
            imageurl = '{}/{}'.format(self.url, app['image'])
            self.storeImage(imageurl, app['id'])
        return ans.json()

    def storeImage(self, imageurl, id):
        image = requests.get(imageurl, stream=True)

        image.raw.decode_content = True
        if not os.path.exists(CACHE_DIR):
            os.mkdir(CACHE_DIR)

        with open(CACHE_DIR + str(id), 'wb') as target:
            shutil.copyfileobj(image.raw, target)
            print("written to: " + CACHE_DIR + str(id))

    @staticmethod
    def instanciate():
        configuration = Configuration()
        preferences = configuration.get('preferences')

        https_protocol = preferences['https_protocol']
        base_url = preferences['base_url']
        client_token = preferences['client_token']

        cache = Cache(https_protocol,
                      base_url,
                      client_token)

        cache.startCaching()


src/client.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# This file is part of gotify-indicator
#
# Copyright (c) 2020 Lorenzo Carbonell Cerezo <a.k.a. atareao>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import websocket
import time
import requests
import threading
import json

# also defined in indicator.py
SOCKET_ACTIVE = 1
SOCKET_INACTIVE = 0
SOCKET_ERROR = -1

class GotifyClient(threading.Thread):
    def __init__(self, https_protocol, base_url, application_name,
                 application_token, client_token, on_message, debug=False):
        threading.Thread.__init__(self)
        protocol = 'https' if https_protocol else 'http'
        ws_protocol = 'wss' if https_protocol else 'ws'
        self.url = '{}://{}'.format(protocol, base_url)
        self.wss_url = '{}://{}/stream'.format(ws_protocol, base_url)
        self.application_name = application_name
        self.application_token = application_token
        self.debug = debug
        self.set_icon = None
        self.running = False
        self.error = False
        self.notstopped = True
        self.ws = websocket.WebSocketApp(self.wss_url,
                                         on_message=on_message,
                                         on_error=self.on_error,
                                         on_close=self.on_close,
                                         header={"X-Gotify-Key":
                                                 client_token})
        if debug:
            websocket.enableTrace(True)
        self.set_running(False)

    def run(self):
        if self.debug:
            print("Start Messagethread")
        self.set_running(True)
        self.ws.run_forever()

        while self.notstopped:
            time.sleep(60)
            if self.debug:
                print("Is Messagethread running: "+str(self.running))

            self.set_running(True)
            self.ws.run_forever()
            self.set_running(False)

    def set_icon_callback(self, seticonfunction):
        self.set_icon = seticonfunction

    def close(self):
        self.ws.close()
        self.set_running(False)
        self.notstopped = False

    def set_running(self, running):
        self.running = running
        if self.set_icon is not None:
            if self.running:
                self.set_icon(SOCKET_ACTIVE)
                self.error = False
            else:
                if self.error:
                    self.set_icon(SOCKET_ERROR)
                else:
                    self.set_icon(SOCKET_INACTIVE)


    def is_running(self):
        return self.running

    def on_message(self, message):
        print(message)
        message = json.loads(message)
        if self.application_name != message['title']:
            print(message['title'])
            print(message)

    def on_error(self, error):
        print(error)
        self.error = True

    def on_close(self):
        self.close()
        print("### closed ###")

    def get_health(self):
        headers = {'X-Gotify-Key': self.application_token}
        url = '{}/{}'.format(self.url, 'health')
        ans = requests.get(url, headers=headers)
        return ans.json()

    def send_message(self, message, title=None, priority=0):
        headers = {'X-Gotify-Key': self.application_token}
        json = {'message': message,
                'priority': priority}
        if title:
            json['title'] = title
        url = '{}/{}'.format(self.url, 'message')
        ans = requests.post(url, headers=headers, json=json)
        return ans.json()


if __name__ == "__main__":
    import time
    from configurator import Configuration
    configuration = Configuration()
    preferences = configuration.get('preferences')

    https_protocol = preferences['https_protocol']
    base_url = preferences['base_url']
    application_name = preferences['application_name']
    application_token = preferences['application_token']
    client_token = preferences['client_token']

    def on_message(ws, message):
        print(message)

    try:
        gc = GotifyClient(https_protocol,
                          base_url,
                          application_name,
                          application_token,
                          client_token,
                          on_message,
                          True)
        print(gc.get_health())
        gc.start()

        while True:
            time.sleep(5)
            gc.send_message("Ejemplo")

    except KeyboardInterrupt as e:
        gc.close()
        print(e)


src/config.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# This file is part of gotify-indicator
#
# Copyright (c) 2020 Lorenzo Carbonell Cerezo <a.k.a. atareao>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import sys
import os
import locale
import gettext

PARAMS = {'stats': {},
          'preferences': {'theme-light': True,
                          'https_protocol': False,
                          'base_url': '',
                          'application_name': '',
                          'application_token': '',
                          'client_token': '',
                          'notification_sound': '',
                          'appid': ''}
          }

CONFIG_DIR = os.path.join(os.path.expanduser('~'), '.config/gotify-indicator')
CONFIG_FILE = os.path.join(CONFIG_DIR, 'gotify-indicator.conf')
CACHE_DIR = os.path.join(os.path.expanduser('~'), '.cache/gotify-indicator/')


def is_package():
    return __file__.find('src') < 0


APP = 'gotify-indicator'
APPNAME = 'Gotify Indicator'

# check if running from source
if is_package():
    ROOTDIR = '/opt/extras.ubuntu.com/gotify-indicator/share'
    LANGDIR = os.path.join(ROOTDIR, 'locale-langpack')
    APPDIR = os.path.join(ROOTDIR, APP)
    CHANGELOG = os.path.join(APPDIR, 'changelog')
    ICONDIR = os.path.join(ROOTDIR, 'icons')
    SOUNDDIR = os.path.join(ROOTDIR, 'sounds')
    HTML_GRAPH = os.path.join(APPDIR, 'graph', 'graph.html')
    AUTOSTARTDIR = os.path.join(ROOTDIR, 'autostart')
else:
    ROOTDIR = os.path.abspath(os.path.dirname(__file__))
    LANGDIR = os.path.normpath(os.path.join(ROOTDIR, '../template1'))
    APPDIR = ROOTDIR
    DEBIANDIR = os.path.normpath(os.path.join(ROOTDIR, '../debian'))
    CHANGELOG = os.path.join(DEBIANDIR, 'changelog')
    ICONDIR = os.path.normpath(os.path.join(ROOTDIR, '../data/icons/'))
    SOUNDDIR = os.path.normpath(os.path.join(ROOTDIR, '../data/sounds/'))
    HTML_GRAPH = os.path.join(ROOTDIR, 'graph', 'graph.html')
    AUTOSTARTDIR = os.path.normpath(
        os.path.join(ROOTDIR, '../data/autostart/'))
AUTOSTART = os.path.join(AUTOSTARTDIR, 'gotify-indicator-autostart.desktop')
ICON = os.path.join(ICONDIR, 'gotify-indicator.svg')
ICON_ACTIVED_LIGHT = os.path.join(ICONDIR, 'gotify-indicator-active.svg')
ICON_PAUSED_LIGHT = os.path.join(ICONDIR, 'gotify-indicator-paused.svg')
ICON_ACTIVED_DARK = os.path.join(ICONDIR, 'gotify-indicator-active-dark.svg')
ICON_PAUSED_DARK = os.path.join(ICONDIR, 'gotify-indicator-paused-dark.svg')

f = open(CHANGELOG, 'r')
line = f.readline()
f.close()
pos = line.find('(')
posf = line.find(')', pos)
VERSION = line[pos + 1:posf].strip()
if not is_package():
    VERSION = VERSION + '-src'

try:
    current_locale, encoding = locale.getdefaultlocale()
    language = gettext.translation(APP, LANGDIR, [current_locale])
    language.install()
    print(language)
    if sys.version_info[0] == 3:
        _ = language.gettext
    else:
        _ = language.ugettext
except Exception as e:
    print(e)
    _ = str
APPNAME = _(APPNAME)


src/basedialog.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# This file is part of gotify-indicator
#
# Copyright (c) 2020 Lorenzo Carbonell Cerezo <a.k.a. atareao>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import gi
try:
    gi.require_version('Gtk', '3.0')
    gi.require_version('Gdk', '3.0')
except ValueError as e:
    print(e)
    exit(1)
from gi.repository import Gdk, Gtk
import config


class BaseDialog(Gtk.Dialog):
    def __init__(self, title, window=None, ok_button=True, cancel_button=True):
        Gtk.Dialog.__init__(self, title, window)
        self.set_modal(True)
        self.set_destroy_with_parent(True)
        if ok_button:
            self.add_button(Gtk.STOCK_OK, Gtk.ResponseType.ACCEPT)
        if cancel_button:
            self.add_button(Gtk.STOCK_CANCEL, Gtk.ResponseType.CANCEL)
        self.set_default_response(Gtk.ResponseType.ACCEPT)
        self.set_resizable(False)
        self.set_icon_from_file(config.ICON)
        self.connect('realize', self.on_realize)
        self.init_ui()
        self.show_all()

    def init_ui(self):
        vbox0 = Gtk.VBox(spacing=5)
        vbox0.set_border_width(5)
        self.get_content_area().add(vbox0)

        frame1 = Gtk.Frame()
        vbox0.add(frame1)

        self.grid = Gtk.Grid()
        self.grid.set_row_spacing(10)
        self.grid.set_column_spacing(10)
        self.grid.set_margin_bottom(10)
        self.grid.set_margin_start(10)
        self.grid.set_margin_end(10)
        self.grid.set_margin_top(10)
        frame1.add(self.grid)

    def on_realize(self, *_):
        monitor = Gdk.Display.get_primary_monitor(Gdk.Display.get_default())
        scale = monitor.get_scale_factor()
        monitor_width = monitor.get_geometry().width / scale
        monitor_height = monitor.get_geometry().height / scale
        width = self.get_preferred_width()[0]
        height = self.get_preferred_height()[0]
        self.move((monitor_width - width)/2, (monitor_height - height)/2)


if __name__ == '__main__':
    dialog = BaseDialog('Test')
    dialog.run()


src/indicator.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# This file is part of gotify-indicator
#
# Copyright (c) 2020 Lorenzo Carbonell Cerezo <a.k.a. atareao>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import sys
import gi
try:
    gi.require_version('Gtk', '3.0')
    gi.require_version('Gdk', '3.0')
    gi.require_version('AppIndicator3', '0.1')
    gi.require_version('GdkPixbuf', '2.0')
    gi.require_version('Notify', '0.7')
except Exception as e:
    print(e)
    exit(-1)
from gi.repository import Gtk
from gi.repository import Gdk
from gi.repository import GLib
from gi.repository import AppIndicator3
from gi.repository import GdkPixbuf
from gi.repository import Notify
import os
import json
import webbrowser
import dbus
from config import _
from preferences import Preferences
import config
from configurator import Configuration
from client import GotifyClient
from message_dialog import MessageDialog
from cache import Cache
from pydub import AudioSegment
from pydub.playback import play


class Indicator(object):

    def __init__(self):

        self.indicator = AppIndicator3.Indicator.new(
            'gotify-indicator',
            'gotify-indicator',
            AppIndicator3.IndicatorCategory.APPLICATION_STATUS)
        self.indicator.set_menu(self.build_menu())
        self.indicator.set_label('', '')
        self.indicator.set_status(AppIndicator3.IndicatorStatus.ACTIVE)
        self.notification = Notify.Notification.new('', '', None)
        self.gotify_client = None
        self.load_preferences()

    def set_icon(self, active=True):
        if active:
            if self.theme_light:
                icon = config.ICON_ACTIVED_LIGHT
            else:
                icon = config.ICON_ACTIVED_DARK
        else:
            if self.theme_light:
                icon = config.ICON_PAUSED_LIGHT
            else:
                icon = config.ICON_PAUSED_DARK
        self.indicator.set_icon(icon)

    def load_preferences(self):
        configuration = Configuration()
        preferences = configuration.get('preferences')
        self.theme_light = preferences['theme-light']
        self.application_name = preferences['application_name']

        self.stop()
        self.start()

    def build_menu(self):
        menu = Gtk.Menu()
        self.menu_toggle_service = Gtk.MenuItem.new_with_label(
            _('Start service'))
        self.menu_toggle_service.connect('activate', self.toggle_service)
        menu.append(self.menu_toggle_service)

        menu.append(Gtk.SeparatorMenuItem())

        menu_send_message = Gtk.MenuItem.new_with_label(
            _('Send message'))
        menu_send_message.connect('activate', self.on_send_message)
        menu.append(menu_send_message)

        menu.append(Gtk.SeparatorMenuItem())

        menu_preferences = Gtk.MenuItem.new_with_label(_('Preferences'))
        menu_preferences.connect('activate', self.show_preferences)
        menu.append(menu_preferences)

        menus_help = Gtk.MenuItem.new_with_label(_('Help'))
        menus_help.set_submenu(self.get_help_menu())
        menu.append(menus_help)

        menu.append(Gtk.SeparatorMenuItem())

        menu_quit = Gtk.MenuItem. new_with_label(_('Quit'))
        menu_quit.connect('activate', self.quit)
        menu.append(menu_quit)
        menu.show_all()
        return menu

    def on_send_message(self, widget):
        widget.set_sensitive(False)
        messageDialog = MessageDialog()
        response = messageDialog.run()
        if response == Gtk.ResponseType.ACCEPT:
            text = messageDialog.get_text()
            if text != '':
                self.gotify_client.send_message(text)
        messageDialog.destroy()
        widget.set_sensitive(True)

    def show_preferences(self, widget):
        widget.set_sensitive(False)
        preferences_dialog = Preferences()
        response = preferences_dialog.run()
        if response == Gtk.ResponseType.ACCEPT:
            preferences_dialog.save()
            configuration = Configuration()
            preferences = configuration.get('preferences')
            self.theme_light = preferences['theme-light']
            self.application_name = preferences['application_name']

            if self.gotify_client is not None and \
                    self.gotify_client.is_running():
                self.stop()
                self.start()
        preferences_dialog.destroy()
        widget.set_sensitive(True)

    def get_help_menu(self):
        help_menu = Gtk.Menu()

        homepage_item = Gtk.MenuItem.new_with_label(_('Homepage'))
        homepage_item.connect(
            'activate',
            lambda x: webbrowser.open(
                'http://www.atareao.es/apps/gotify-indicator/'))
        help_menu.append(homepage_item)

        help_item = Gtk.MenuItem.new_with_label(_('Get help online...'))
        help_item.connect(
            'activate',
            lambda x: webbrowser.open(
                'http://www.atareao.es/apps/gotify-indicator/'))
        help_menu.append(help_item)

        translate_item = Gtk.MenuItem.new_with_label(_(
            'Translate this application...'))
        translate_item.connect(
            'activate',
            lambda x: webbrowser.open(
                'http://www.atareao.es/apps/gotify-indicator/'))
        help_menu.append(translate_item)

        bug_item = Gtk.MenuItem.new_with_label(_('Report a bug...'))
        bug_item.connect(
            'activate',
            lambda x: webbrowser.open('https://github.com/atareao\
/gotify-indicator/issues'))
        help_menu.append(bug_item)

        help_menu.append(Gtk.SeparatorMenuItem())

        twitter_item = Gtk.MenuItem.new_with_label(_('Found me in Twitter'))
        twitter_item.connect(
            'activate',
            lambda x: webbrowser.open('https://twitter.com/atareao'))
        help_menu.append(twitter_item)
        #
        github_item = Gtk.MenuItem.new_with_label(_('Found me in GitHub'))
        github_item.connect(
            'activate',
            lambda x: webbrowser.open('https://github.com/atareao'))
        help_menu.append(github_item)

        mastodon_item = Gtk.MenuItem.new_with_label(_('Found me in Mastodon'))
        mastodon_item.connect(
            'activate',
            lambda x: webbrowser.open('https://mastodon.social/@atareao'))
        help_menu.append(mastodon_item)

        about_item = Gtk.MenuItem.new_with_label(_('About'))
        about_item.connect('activate', self.menu_about_response)

        help_menu.append(Gtk.SeparatorMenuItem())

        help_menu.append(about_item)
        return help_menu

    def menu_about_response(self, widget):
        widget.set_sensitive(False)
        ad = Gtk.AboutDialog()
        ad.set_name(config.APPNAME)
        ad.set_version(config.VERSION)
        ad.set_copyright('Copyrignt (c) 2020\nLorenzo Carbonell')
        ad.set_comments(_('Gotify Indicator'))
        ad.set_license('''
Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.''')
        ad.set_website('')
        ad.set_website_label('http://www.atareao.es')
        ad.set_authors(['Lorenzo Carbonell Cerezo <a.k.a. atareao>'])
        ad.set_translator_credits('Lorenzo Carbonell Cerezo <a.k.a. atareao>')
        ad.set_documenters(['Lorenzo Carbonell Cerezo <a.k.a. atareao>'])
        ad.set_artists(['Freepik <https://www.flaticon.com/authors/freepik>'])
        ad.set_logo(GdkPixbuf.Pixbuf.new_from_file(config.ICON))
        ad.set_icon(GdkPixbuf.Pixbuf.new_from_file(config.ICON))
        ad.set_program_name(config.APPNAME)

        monitor = Gdk.Display.get_primary_monitor(Gdk.Display.get_default())
        scale = monitor.get_scale_factor()
        monitor_width = monitor.get_geometry().width / scale
        monitor_height = monitor.get_geometry().height / scale
        width = ad.get_preferred_width()[0]
        height = ad.get_preferred_height()[0]
        ad.move((monitor_width - width)/2, (monitor_height - height)/2)

        ad.run()
        ad.destroy()
        widget.set_sensitive(True)

    def toggle_service(self, widget):
        if self.gotify_client is None:
            self.start()
        else:
            self.stop()

    def stop(self):
        self.set_icon(False)
        if self.gotify_client is not None:
            if self.gotify_client.is_running():
                self.gotify_client.close()
        self.gotify_client = None
        self.menu_toggle_service.set_label(_('Start service'))

    def on_message(self, message):
        message = json.loads(message)
        icon = os.path.join(config.CACHE_DIR, str(message['appid']))
        if message['title'] != self.application_name:
            self.notification.update(message['title'],
                                     message['message'],
                                     icon)
            self.notification.show()
            try:
                sound = self.getNotificationSound()
                if sound.strip():
                    play(AudioSegment.from_mp3(sound))
            except Exception as e:
                print("Exception while playing sound: "+str(e))

    def getNotificationSound(self):
        path = ""  # os.path.join(config.SOUNDDIR, 'default.mp3')
        configuration = Configuration()
        preferences = configuration.get('preferences')
        try:
            custom_path = preferences['notification_sound']
            if custom_path:
                path = custom_path
        except:
            print("No setting for notificationsound")

        return path

    def start(self):
        configuration = Configuration()
        preferences = configuration.get('preferences')

        https_protocol = preferences['https_protocol']
        base_url = preferences['base_url']
        application_name = preferences['application_name']
        application_token = preferences['application_token']
        client_token = preferences['client_token']

        if base_url and application_name and application_token \
                and client_token:
            self.gotify_client = GotifyClient(https_protocol,
                                              base_url,
                                              application_name,
                                              application_token,
                                              client_token,
                                              self.on_message,
                                              True)
            self.gotify_client.start()
            self.set_icon(True)
            self.menu_toggle_service.set_label(_('Stop service'))
            return True
        message = _('Please configure Gotify Indicator')
        icon = os.path.join(config.ICONDIR, 'gotify-indicator.svg')
        self.notification.update('Gofify-Indictator',
                                 message,
                                 icon)
        self.notification.show()

    def quit(self, menu_item):
        self.stop()
        Gtk.main_quit()
        # If Gtk throws an error or just a warning, main_quit() might not
        # actually close the app
        sys.exit(0)


def main():
    if dbus.SessionBus().request_name(
        'es.atareao.GotifyIndicator') !=\
            dbus.bus.REQUEST_NAME_REPLY_PRIMARY_OWNER:
        print("application already running")
        exit(0)

    try:
        Cache.instanciate()
    except Exception as e:
        print('Could not cache images!'+str(e))

    Notify.init('Gotify')
    Indicator()
    Gtk.main()

if __name__ == '__main__':
    main()
